src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -4254,70 +4150,65 @@ int putIntCompact(long intCompact) {
      *         {@code BigDecimal}
      */
     private String layoutChars(boolean sci) {
+        int scale = this.scale;
+        long intCompact = this.intCompact;
         if (scale == 0)                      // zero scale is trivial
-            return (intCompact != INFLATED) ?
-                Long.toString(intCompact):
-                intVal.toString();
+            return unscaledString();
         if (scale == 2  &&
             intCompact >= 0 && intCompact < Integer.MAX_VALUE) {
             // currency fast path
-            int lowInt = (int)intCompact % 100;
             int highInt = (int)intCompact / 100;
-            return (Integer.toString(highInt) + '.' +
-                    StringBuilderHelper.DIGIT_TENS[lowInt] +
-                    StringBuilderHelper.DIGIT_ONES[lowInt]) ;
+            short pair = DecimalDigits.digitPair((int)intCompact % 100);
+            return "" + highInt + '.' + (char)(pair & 0xff) + (char)(pair >> 8);
```

Comment:
FTR: while this works (since `pair` is never negative), casting `pair` to `char` or changing to `(char)((pair >> 8) & 0xff)` would be necessary for the general case.